// sanity-plugin-seo-tools/lib/cjs/utils/prepareRemoteData.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "prepareRemoteData", {
    enumerable: true,
    get: function() {
        return prepareRemoteData;
    }
});
var _constants = require("../constants");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _asyncToGenerator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
var __generator = (void 0) && (void 0).__generator || function(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
};
function prepareRemoteData(url, options) {
    return _prepareRemoteData.apply(this, arguments);
}
function _prepareRemoteData() {
    _prepareRemoteData = _asyncToGenerator(function(url, options) {
        var _doc_querySelector, _doc_querySelector1, _doc_querySelector2, _doc_querySelector3, _options_contentSelector, contentSelector, result, html, parser, doc, _doc_documentElement_lang, langCulture, _LangCultureMap_langCulture_toLowerCase, title, description, content;
        return __generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    contentSelector = (_options_contentSelector = options.contentSelector) !== null && _options_contentSelector !== void 0 ? _options_contentSelector : "body";
                    return [
                        4,
                        fetch(url.toString())
                    ];
                case 1:
                    result = _state.sent();
                    return [
                        4,
                        result.text()
                    ];
                case 2:
                    html = _state.sent();
                    parser = new DOMParser();
                    doc = parser.parseFromString(html, "text/html");
                    // remove all script tags
                    Array.from(doc.body.querySelectorAll("script")).forEach(function(script) {
                        script.remove();
                    });
                    langCulture = (_doc_documentElement_lang = doc.documentElement.lang) !== null && _doc_documentElement_lang !== void 0 ? _doc_documentElement_lang : _constants.LangCultureMap.en;
                    if (!langCulture.includes("-")) {
                        ;
                        langCulture = (_LangCultureMap_langCulture_toLowerCase = _constants.LangCultureMap === null || _constants.LangCultureMap === void 0 ? void 0 : _constants.LangCultureMap[langCulture.toLowerCase()]) !== null && _LangCultureMap_langCulture_toLowerCase !== void 0 ? _LangCultureMap_langCulture_toLowerCase : _constants.LangCultureMap.en;
                    }
                    title = doc.title || ((_doc_querySelector = doc.querySelector("h1")) === null || _doc_querySelector === void 0 ? void 0 : _doc_querySelector.textContent) || ((_doc_querySelector1 = doc.querySelector("h2")) === null || _doc_querySelector1 === void 0 ? void 0 : _doc_querySelector1.textContent) || "";
                    description = ((_doc_querySelector2 = doc.querySelector('meta[name="description"]')) === null || _doc_querySelector2 === void 0 ? void 0 : _doc_querySelector2.content) || "";
                    content = ((_doc_querySelector3 = doc.querySelector(contentSelector)) === null || _doc_querySelector3 === void 0 ? void 0 : _doc_querySelector3.innerHTML) || "";
                    return [
                        2,
                        {
                            locale: langCulture,
                            title: title,
                            description: description,
                            content: content
                        }
                    ];
            }
        });
    });
    return _prepareRemoteData.apply(this, arguments);
}
