// sanity-plugin-seo-tools/lib/cjs/components/SeoToolsPane.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "SeoToolsPane", {
    enumerable: true,
    get: function() {
        return SeoToolsPane;
    }
});
var _react = /*#__PURE__*/ _interopRequireWildcard(require("react"));
var _context = require("../context");
var _seoToolsPaneView = require("./SeoToolsPaneView");
var _seoToolsPaneEmptyView = require("./SeoToolsPaneEmptyView");
function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interopRequireWildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {};
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function _objectSpread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _defineProperty(target, key, source[key]);
        });
    }
    return target;
}
var SeoToolsPane = function(param) {
    var options = param.options, document = param.document;
    var _document_displayed, _document_displayed1;
    var contextValue = (0, _react.useMemo)(function() {
        return _objectSpread({
            fetch: true,
            select: function() {
                return {};
            },
            render: function(seo, data, serpPreview) {
                return serpPreview;
            },
            resolveProductionUrl: function() {
                return new URL("https://sanity.io");
            }
        }, options);
    }, []);
    return /*#__PURE__*/ _react.default.createElement(_context.SeoToolsContext.Provider, {
        value: contextValue
    }, (document === null || document === void 0 ? void 0 : (_document_displayed = document.displayed) === null || _document_displayed === void 0 ? void 0 : _document_displayed._rev) && /*#__PURE__*/ _react.default.createElement(_seoToolsPaneView.SeoToolsPaneView, {
        document: document === null || document === void 0 ? void 0 : document.displayed
    }), !(document === null || document === void 0 ? void 0 : (_document_displayed1 = document.displayed) === null || _document_displayed1 === void 0 ? void 0 : _document_displayed1._rev) && /*#__PURE__*/ _react.default.createElement(_seoToolsPaneEmptyView.SeoToolsPaneEmptyView, null));
};
